### PR TITLE
Add `markdownlint` and dev Ruff to `pre-commit`

### PR DESCRIPTION
**Pull request by JonathanPlasse**
Hi,
I propose adding _markdownlint_ and _Prettier_ to pre-commit.
I already added _markdownlint_ but I did not add _Prettier_ because `dev-generate-all` is incompatible with it.
We would have to run _Prettier_ inside `dev-generate-all`.
To do this we either require the developer to have node installed on their machine or use [nodeenv](https://github.com/ekalinin/nodeenv) which only requires _Python_ to create a _Node_ virtual environment where we would run _Prettier._

_Prettier_ would be really useful to format `ruff.schema.json` as it is required by `schemastore.org`.

Also maybe we should consider using for the ruff hook `cargo run` with `repo: local` instead of `ruff-pre-commit` which sometimes fails because the version for the latest release is not out yet.

### CODE REVIEW

File: CODE_OF_CONDUCT.md [hunk 1]
```diff
@@ -17,23 +17,23 @@ diverse, inclusive, and healthy community.
...
 community include:
 
-* Demonstrating empathy and kindness toward other people
-* Being respectful of differing opinions, viewpoints, and experiences
-* Giving and gracefully accepting constructive feedback
-* Accepting responsibility and apologizing to those affected by our mistakes,
+- Demonstrating empathy and kindness toward other people
+- Being respectful of differing opinions, viewpoints, and experiences
+- Giving and gracefully accepting constructive feedback
+- Accepting responsibility and apologizing to those affected by our mistakes,
```
Comment (not-my-profile) [inline CODE_OF_CONDUCT.md:23]: It's a nit but I prefer `*` for bullet lists over `-`.

File: python/ruff/__main__.py [hunk 2]
```diff
@@ -1,7 +1,8 @@
 import os
 import sys
 import sysconfig
+from pathlib import Path
 
 if __name__ == "__main__":
-    ruff = os.path.join(sysconfig.get_path("scripts"), "ruff")
+    ruff = Path(sysconfig.get_path("scripts")) / "ruff"
```
Comment (not-my-profile) [inline python/ruff/__main__.py:7]: Is this change related to prettier / markdownlint?

Comment (JonathanPlasse) [reply python/ruff/__main__.py:7]: No, but it was failing the Ruff hook. So I fixed it.

**Comment by sbrugman** [comment]
Cargo format is part of our GitHub actions pipeline. If markdownlint is added, any reason to no enforce it there?

**Comment by JonathanPlasse** [comment]
We could use [pre-commit.ci](https://pre-commit.ci/) that would run on pull-request to do this.

**Comment by sbrugman** [comment]
> We could use [pre-commit.ci](https://pre-commit.ci/) that would run on pull-request to do this.

Certainly an option. There is also a markdownlint github action: 
https://github.com/actionshub/markdownlint (that does not require to install a github application)

File: CODE_OF_CONDUCT.md (hunk 1 above)
Comment (sbrugman) [reply CODE_OF_CONDUCT.md:23]: Markdownlint supports the asterix style: https://github.com/markdownlint/markdownlint/blob/main/docs/RULES.md#md004---unordered-list-style

**Comment by charliermarsh** [comment]
> I already added markdownlint but I did not add Prettier because dev-generate-all is incompatible with it.

Can you expand on this? How is it incompatible?

**Comment by JonathanPlasse** [comment]
Dev generate all will generate the different file which will be then formated by prettier.
Pre-commit will always fail at the one of the hook as the output of prettier is different from dev generate all.
For it to be compatible it would need that dev generate all generate the different files already well formated.

**Comment by not-my-profile** [comment]
>To do this we either require the developer to have node installed on their machine or use [nodeenv](https://github.com/ekalinin/nodeenv) which only requires Python to create a Node virtual environment where we would run Prettier.

I'd really rather not add Node as a dev dependency.

> Prettier would be really useful to format ruff.schema.json as it is required by schemastore.org.

I think the better approach would be to implement https://github.com/SchemaStore/schemastore/issues/2731 then our schema file would no longer have to be part of the schemastore repository.

**Comment by charliermarsh** [comment]
Yeah I'd like to avoid adding Node as a dependency. I'm happy to just do this manually from time to time, even though it's not a great practice.

**Comment by JonathanPlasse** [comment]
I added prettier in pre-commit.
I had to:
- Ignore formatting for the table (c.f. prettier/prettier-vscode#762)
- Use `-` as bullet as prettier does not support `*` bullet (c.f. prettier/prettier#4440).
  prettier [option philosophy](https://prettier.io/docs/en/option-philosophy.html) is to have few option so that we do not debate on which style to chose.
  > By far the biggest reason for adopting Prettier is to stop all the ongoing debates over styles.
  > Yet the more options Prettier has, the further from the above goal it gets. The debates over styles just turn into debates over which Prettier options to use.

File: CODE_OF_CONDUCT.md (hunk 1 above)
Comment (JonathanPlasse) [reply CODE_OF_CONDUCT.md:23]: Prettier does not support it.

**Comment by charliermarsh** [comment]
> I added prettier in pre-commit.

Does this require that the user have Node installed? Or how does it work?

(I don't feel strongly about those individual style decisions, like `*` vs. `-`. Even if I have preferences, it's not a blocker for me.)

**Comment by JonathanPlasse** [comment]
> > I added prettier in pre-commit.
> 
> Does this require that the user have Node installed? Or how does it work?

No, it does not require Node. If pre-commit is installed it will run the prettier hook with `nodeenv` but the developer only need to have pre-commit installed and nothing else.
For the CI, I would suggest using [pre-commit.ci](https://pre-commit.ci) to run pre-commit. To avoid checking twice `fmt`, `clippy`, and `dev-generate-all`, we can edit `.pre-commit-config.yaml` to skip them when in CI.

**Comment by charliermarsh** [comment]
Candidly, while I definitely appreciate the motivation, I'm probably not going to enable pre-commit CI right now, since I don't use pre-commit personally!

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -1,16 +1,25 @@
...
+
+  - repo: https://github.com/igorshubovych/markdownlint-cli
+    rev: v0.33.0
+    hooks:
+      - id: markdownlint-fix
+        args:
+          - --disable
+          - MD013
+          - MD033
+          - --
```
Comment (charliermarsh) [inline .pre-commit-config.yaml:21]: Should this now be removed? (Are we using Prettier instead?)

Comment (JonathanPlasse) [reply .pre-commit-config.yaml:21]: Prettier is only a formatter, it will not catch some lint that markdownlint has.
e.g. [MD001](https://github.com/DavidAnson/markdownlint/blob/main/doc/md001.md) heading-increment/header-increment - Heading levels should only increment by one level at a time

Comment (JonathanPlasse) [reply .pre-commit-config.yaml:21]: We should keep both.

Comment (charliermarsh) [reply .pre-commit-config.yaml:21]: I see, does markdownlint do any autofixing? Would the tools ever conflict?

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -1,16 +1,25 @@
...
   - repo: https://github.com/abravalheri/validate-pyproject
     rev: v0.10.1
     hooks:
       - id: validate-pyproject
 
+  - repo: https://github.com/pre-commit/mirrors-prettier
+    rev: v3.0.0-alpha.4
+    hooks:
+      - id: prettier
+        exclude: ^ruff.schema.json$
```
[the remark this answers is not in the document]
Comment (JonathanPlasse) [reply .pre-commit-config.yaml:11]: Be cause `schemars` formatting conflicts with prettier.
I think it was not the case before.
Let me do a bisect to check if it is the case.

[the remark this answers is not in the document]
Comment (JonathanPlasse) [reply .pre-commit-config.yaml:11]: Generated from `schemars`:
```json
"properties": {
    "allowed-confusables": {
      "description": "A list of allowed \"confusable\" Unicode characters to ignore when enforcing `RUF001`, `RUF002`, and `RUF003`.",
      "type": [
        "array",
        "null"
      ],
```
After Prettier formatting:
```json
"properties": {
    "allowed-confusables": {
      "description": "A list of allowed \"confusable\" Unicode characters to ignore when enforcing `RUF001`, `RUF002`, and `RUF003`.",
      "type": ["array", "null"],
```

[the remark this answers is not in the document]
Comment (JonathanPlasse) [reply .pre-commit-config.yaml:11]: So I checked, `ruff.schema.json` was never well formatted.

[the remark this answers is not in the document]
Comment (JonathanPlasse) [reply .pre-commit-config.yaml:11]: We could do it in a future pull request.

File: CODE_OF_CONDUCT.md (hunk 1 above)
Comment (not-my-profile) [reply CODE_OF_CONDUCT.md:23]: I'd rather drop prettier than change our formatting just to cater to prettier.

Comment (not-my-profile) [reply CODE_OF_CONDUCT.md:23]: Doesn't pre-commit download and install node behind the scenes?

Comment (JonathanPlasse) [reply CODE_OF_CONDUCT.md:23]: Yes, it does but I do not think it is a burden on the developer experience.

Comment (not-my-profile) [reply CODE_OF_CONDUCT.md:23]: Having to install pre-commit is an additional step and thereby somewhat of a burden. Personally I rather avoid tools that download stuff behind the scenes.

Comment (not-my-profile) [reply CODE_OF_CONDUCT.md:23]: And the drawbacks of depending on an entire other ecosystem don't go away just because you install it automatically.

Comment (charliermarsh) [reply CODE_OF_CONDUCT.md:23]: Yeah I've given it some thought... and I think I'd prefer not to add Prettier as a dev dependency. I'm happy to run Prettier as a one-time thing over the YAML files in this PR, and also to clean up the Markdown a bit (you fixed some good things in the README like trailing space, but we could roll back the `*` changes and such), but I'd rather not add Prettier on an ongoing basis.

**Comment by charliermarsh** [comment]
@JonathanPlasse - I'd like to pull in some of these changes, but exclude Prettier for now. Would you like me to do it in a separate PR, or push to this branch?

**Comment by JonathanPlasse** [comment]
Hi, sorry for the late response.
I rebase and remove prettier.
